File: web/src/types/series.ts
/**
 * Wire types for the time-range query API. Mirrors app/schemas/series.py.
 *
 * A point is the same shape whether it came from raw 1-second samples or from
 * an hour-wide rollup — that is deliberate, and it is why `source` and
 * `resolution_seconds` travel with the response. A chart that does not say
 * which of the two it is drawing is making a claim it cannot support.
 */

export type SeriesSource = "raw" | "1m" | "5m" | "1h"

export type SeriesDomain = "system" | "disk_usage" | "disk_io" | "net" | "latency"

interface Bucket {
  ts: string
  /** Total measured seconds behind this point. */
  sample_weight: number
  /** Original samples behind it — 1 and 6 are both "a minute of CPU". */
  sample_count: number
}

export interface SystemPoint extends Bucket {
  cpu_percent: number | null
  cpu_percent_min: number | null
  cpu_percent_max: number | null
  cpu_user_percent: number | null
  cpu_system_percent: number | null
  cpu_iowait_percent: number | null
  ctx_switches_per_s: number | null
  load1: number | null
  load5: number | null
  load15: number | null
  mem_total_bytes: number | null
  mem_used_bytes: number | null
  mem_available_bytes: number | null
  mem_percent: number | null
  mem_percent_min: number | null
  mem_percent_max: number | null
  swap_percent: number | null
  process_count: number | null
  active_connections: number | null
  temperature_celsius: number | null
  temperature_celsius_max: number | null
  battery_percent: number | null
}

export interface DiskUsagePoint extends Bucket {
  mount: string
  filesystem: string | null
  total_bytes: number | null
  used_bytes: number | null
  free_bytes: number | null
  percent: number | null
  percent_max: number | null
}

export interface DiskIoPoint extends Bucket {
  disk: string
  read_bytes_per_s: number | null
  write_bytes_per_s: number | null
  read_bytes_per_s_max: number | null
  write_bytes_per_s_max: number | null
  read_iops: number | null
  write_iops: number | null
  busy_percent: number | null
}

export interface NetPoint extends Bucket {
  nic: string
  tx_bytes_per_s: number | null
  rx_bytes_per_s: number | null
  tx_bytes_per_s_max: number | null
  rx_bytes_per_s_max: number | null
  errors_in_per_s: number | null
  drops_in_per_s: number | null
}

export interface LatencyPoint extends Bucket {
  target: string
  reachable: boolean
  rtt_ms_avg: number | null
  rtt_ms_min: number | null
  rtt_ms_max: number | null
  packet_loss_percent: number | null
  /** Fraction of the bucket the target answered in — trust this over `reachable`. */
  reachable_ratio: number | null
}

export interface Series {
  device_id: string
  start: string
  end: string
  source: SeriesSource
  /** Bucket width the server actually used, which is not the source's own. */
  resolution_seconds: number
  truncated: boolean
  system: SystemPoint[]
  disk_usage: DiskUsagePoint[]
  disk_io: DiskIoPoint[]
  net: NetPoint[]
  latency: LatencyPoint[]
}

File: backend/app/schemas/series.py
"""Wire schemas for the time-range query API.

Each point extends the corresponding agent protocol entry, because a rolled-up
point carries exactly the same columns as a raw one — that identity is enforced
by app/models/rollups.py, where every aggregate is named after the raw column it
summarises. What a rollup adds is the bookkeeping the raw row does not need:
how much measured time went into the bucket, and how many rows.

`sample_count` is not decoration. A bucket built from one sample and a bucket
built from six are both "a minute of CPU", and only this field distinguishes a
quiet machine from one that was disconnected for most of the minute.
"""

from __future__ import annotations

import uuid
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, Field

from app.schemas.protocol import (
    DiskIoEntry,
    DiskUsageEntry,
    LatencyEntry,
    NetEntry,
    SystemSample,
)

SourceName = Literal["raw", "1m", "5m", "1h"]

DomainName = Literal["system", "disk_usage", "disk_io", "net", "latency"]

#: Maps the request's domain name onto the rollup spec's source table. The API
#: never accepts a table name directly — a name outside this mapping is a 422
#: before anything reaches SQL.
DOMAIN_SOURCES: dict[str, str] = {
    "system": "metric_samples",
    "disk_usage": "disk_usage_samples",
    "disk_io": "disk_io_samples",
    "net": "net_samples",
    "latency": "latency_samples",
}


class _Bucket(BaseModel):
    ts: datetime
    #: Total measured seconds behind this point: the sum of the source rows'
    #: `resolution_seconds`. This is the weight every mean above was divided by.
    sample_weight: int
    sample_count: int


class SystemSeriesPoint(SystemSample, _Bucket):
    temperature_celsius_max: float | None = None


class DiskUsageSeriesPoint(DiskUsageEntry, _Bucket):
    percent_max: float | None = None


class DiskIoSeriesPoint(DiskIoEntry, _Bucket):
    read_bytes_per_s_max: float | None = None
    write_bytes_per_s_max: float | None = None
    busy_percent_max: float | None = None


class NetSeriesPoint(NetEntry, _Bucket):
    tx_bytes_per_s_max: float | None = None
    rx_bytes_per_s_max: float | None = None


class LatencySeriesPoint(LatencyEntry, _Bucket):
    #: Fraction of the bucket's measured time in which the target answered.
    #: `reachable` above is `bool_or` — true if it answered even once — so this
    #: is the field to trust for availability.
    reachable_ratio: float | None = None


class SeriesOut(BaseModel):
    """One response covers every requested domain over one window.

    `source` and `resolution_seconds` are part of the payload rather than an
    implementation detail: a chart drawn from 13-minute buckets must be able to
    say so, and "CPU averaged over 13 minutes" is a different claim from "CPU
    every second".
    """

    device_id: uuid.UUID
    start: datetime
    end: datetime
    source: SourceName
    resolution_seconds: int
    #: True when a domain hit the server's row budget and the tail was dropped.
    #: Ask for a narrower window or fewer points rather than trusting the edge.
    truncated: bool = False

    system: list[SystemSeriesPoint] = Field(default_factory=list)
    disk_usage: list[DiskUsageSeriesPoint] = Field(default_factory=list)
    disk_io: list[DiskIoSeriesPoint] = Field(default_factory=list)
    net: list[NetSeriesPoint] = Field(default_factory=list)
    latency: list[LatencySeriesPoint] = Field(default_factory=list)


POINT_MODELS: dict[str, type[BaseModel]] = {
    "system": SystemSeriesPoint,
    "disk_usage": DiskUsageSeriesPoint,
    "disk_io": DiskIoSeriesPoint,
    "net": NetSeriesPoint,
    "latency": LatencySeriesPoint,
}
